CCDB/src/CCDBDownloader.cxx

```diff
@@ -354,42 +354,103 @@ void CCDBDownloader::tryNewHost(PerformData* performData, CURL* easy_handle)
-    performData->locInd++;
+    newLocation = getNewLocation(performData, locations);
   }
 }
 
-void CCDBDownloader::httpRedirect(PerformData* performData, std::string& newUrl, std::string& newLocation, CURL* easy_handle)
+std::string CCDBDownloader::getNewLocation(PerformData* performData, std::vector<std::string>& locations) const
 {
   auto requestData = performData->requestData;
-  newUrl = requestData->hosts.at(performData->hostInd) + newLocation;
-  LOG(debug) << "Trying content location " << newUrl;
-  curl_easy_setopt(easy_handle, CURLOPT_URL, newUrl.c_str());
+  if (performData->locInd < locations.size()) {
+    std::string newLocation = locations.at(performData->locInd++);
+    std::string hostUrl = requestData->hosts.at(performData->hostInd);
+    std::string newUrl = prepareRedirectedURL(newLocation, hostUrl);
+    return newUrl;
+  } else {
+    return "";
+  }
+}
+
+void CCDBDownloader::httpRedirect(PerformData* performData, std::string& newLocation, CURL* easy_handle)
+{
+  auto requestData = performData->requestData;
+  LOG(debug) << "Trying content location " << newLocation;
+  curl_easy_setopt(easy_handle, CURLOPT_URL, newLocation.c_str());
   mHandlesToBeAdded.push_back(easy_handle);
 }
 
 void CCDBDownloader::followRedirect(PerformData* performData, CURL* easy_handle, std::vector<std::string>& locations, bool& rescheduled, bool& contentRetrieved)
 {
-  std::string newLocation = locations.at(performData->locInd++);
-  std::string newUrl;
+  std::string newLocation = getNewLocation(performData, locations);
   if (newLocation.find("alien:/", 0) != std::string::npos || newLocation.find("file:/", 0) != std::string::npos) {
-    getLocalContent(performData, newUrl, newLocation, contentRetrieved, locations);
+    getLocalContent(performData, newLocation, contentRetrieved, locations);
   }
   if (!contentRetrieved && newLocation != "") {
-    httpRedirect(performData, newUrl, newLocation, easy_handle);
+    httpRedirect(performData, newLocation, easy_handle);
     rescheduled = true;
   }
 }
 
+std::string CCDBDownloader::trimHostUrl(std::string full_host_url) const
+{
+  CURLU* host_url = curl_url();
+  curl_url_set(host_url, CURLUPART_URL, full_host_url.c_str(), 0);
+  // Get host part
+  char* host;
+  CURLUcode host_result = curl_url_get(host_url, CURLUPART_HOST, &host, 0);
+  std::string host_name;
+  if (host_result == CURLUE_OK) {
+    // Host part present
+    host_name = host;
+    curl_free(host);
+  } else {
+    LOG(error) << "CCDBDownloader: Malformed url detected when processing redirect, could not identify the host part: " << host;
+    curl_url_cleanup(host_url);
+    return "";
+  }
+  // Get scheme (protocol) part
+  char* scheme;
+  CURLUcode scheme_result = curl_url_get(host_url, CURLUPART_SCHEME, &scheme, 0);
+  curl_url_cleanup(host_url);
+  if (scheme_result == CURLUE_OK) {
+    // If protocol present combine with host
+    curl_free(scheme);
+    return scheme + std::string("://") + host_name;
```

Comment:
@TrifleMichael I'd say using scheme after it's been used is wrong.

Anyway, since this PR I'm no longer able to use condition remapped object from a local ccdb. 

e.g. using something like `--condition-remap "http://localhost:6464=MCH/Calib/BadChannel`

I end up with : 
```
[93683:internal-dpl-ccdb-backend]: [11:50:04][INFO] ccdb reads [localhost:6464/MCH/Calib/BadChannel](http://localhost:6464/MCH/Calib/BadChannel) for 1669594219618 (load to memory, agent_id: nanpc707.in2p3.fr-1699959000-rN3HZ9),
[93683:internal-dpl-ccdb-backend]: [11:50:04][ERROR] Error in fetching object �D�v�m://localhost/MCH/Calib/BadChannel/1669594219618/8b2e5406-82db-11ee-9f12-7f0000010000, curl response code:0
[93683:internal-dpl-ccdb-backend]: [11:50:04][ERROR] File MCH/Calib/BadChannel could not be retrieved. No more hosts to try.
[93683:internal-dpl-ccdb-backend]: [11:50:04][ALARM] Curl request to �D�v�m://localhost/MCH/Calib/BadChannel/1669594219618/8b2e5406-82db-11ee-9f12-7f0000010000, response code: 0
[93683:internal-dpl-ccdb-backend]: [11:50:04][WARN] Did not receive content for MCH/Calib/BadChannel
[93683:internal-dpl-ccdb-backend]:
[93683:internal-dpl-ccdb-backend]: [11:50:04][FATAL] Unable to find object MCH/Calib/BadChannel/1669594219618
```

Note that there's probably two issues : the url name is mangled (probably due to this scheme var being used after free) and the port part is gone...

Reply:
Thank you for bringing this to my attention. Indeed, due to the mistakes the redirected url's can become malformed. I'll start working on fixing that.

Reply:
I made a PR aiming to fix this issue https://github.com/AliceO2Group/AliceO2/pull/12259